### PR TITLE
Split NGINX Prometheus Exporter to a separated container

### PR DESCRIPTION
**Pull request by armandomiani**

### CODE REVIEW

File: 1/README.md [hunk 1]
```diff
@@ -1,406 +0,0 @@
-# <a name="about"></a>About
```
Comment (wgrzelak) [inline 1/README.md]: Why have you deleted it?

Comment (armandomiani) [reply 1/README.md]: My bad. Just re-added.

File: versions.yaml [hunk 2]
```diff
@@ -65,5 +68,24 @@ versions:
...
           version: '1.14.2'
 
+  - dir: 'exporter'
+    repo: nginx-prometheus-exporter0
+    templateSubDir: exporter
+    templateArgs:
+      repo: 'http://nginx.org/packages/mainline/debian/'
+      os: 'stretch'
+    tags:
+      - '0.3.0-debian9'
```
Comment (wgrzelak) [inline versions.yaml]: Please follow the convention. It's good to always take a look at different repos.

File: templates/exporter/Dockerfile.template [hunk 3]
```diff
@@ -0,0 +1,30 @@
+FROM {{ .From }}
+
```
Comment (wgrzelak) [inline templates/exporter/Dockerfile.template]: rm blank line (one of two)

File: templates/exporter/Dockerfile.template [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+{{- $exporter := index .Packages "exporter" }}
+ENV EXPORTER_VERSION {{ $exporter.Version }}
+ENV EXPORTER_SHA256 {{ $exporter.Gpg }}
+
+ENV C2D_RELEASE ${EXPORTER_VERSION}
```
Comment (wgrzelak) [inline templates/exporter/Dockerfile.template]:
```suggestion
ENV C2D_RELEASE {{ $exporter.Version }}
```

File: exporter/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,27 @@
+FROM gcr.io/google-appengine/debian9
+ENV EXPORTER_VERSION 0.3.0
```
Comment (wgrzelak) [inline exporter/Dockerfile:3]:
```suggestion

ENV EXPORTER_VERSION 0.3.0
```

File: 1/README.yaml [hunk 6]
```diff
@@ -170,4 +170,4 @@ volumeReference:
       Contains nginx configuration files, including `nginx.conf`.
       <br><br>
       The default `nginx.conf` include all `.conf` files under the subdirectory
-      `conf.d`.
+      `conf.d`.
```
Comment (wgrzelak) [inline 1/README.yaml]: Can you fix EOF?

**Review by wgrzelak — CHANGES_REQUESTED** [review]

File: versions.yaml (hunk 2 above)
Comment (armandomiani) [reply versions.yaml]: Done. Using consul-docker as a template.

File: templates/exporter/Dockerfile.template (hunk 3 above)
Comment (armandomiani) [reply templates/exporter/Dockerfile.template]: Done.

File: templates/exporter/Dockerfile.template (hunk 4 above)
Comment (armandomiani) [reply templates/exporter/Dockerfile.template]: Remove C2D_RELEASE.

File: 1/README.yaml (hunk 6 above)
Comment (armandomiani) [reply 1/README.yaml]: Done.

File: exporter/Dockerfile (hunk 5 above)
Comment (armandomiani) [reply exporter/Dockerfile:3]: Done

File: versions.yaml [hunk 7]
```diff
@@ -65,5 +68,19 @@ versions:
           gpg: *gpg
           version: '1.14.2'
 
+  - dir: 'exporter'
+    repo: nginx-prometheus-exporter0
```
Comment (wgrzelak) [inline versions.yaml]:
```suggestion
    repo: nginx1
```

We keep the exporter in the same repo.

Comment (armandomiani) [reply versions.yaml]: Fixed.

File: versions.yaml [hunk 8]
```diff
@@ -47,8 +48,10 @@ versions:
         nginx:
           gpg: *gpg
           version: '1.15.12'
+
   - dir: '1/debian9/1.14'
     repo: nginx1
+    templateSubDir: nginx
```
Comment (wgrzelak) [inline versions.yaml]: Can we name the directory more descriptive, to make sure we will not use this template for the future versions?

For example: `nginx-with-exporter`

File: versions.yaml [hunk 9]
```diff
@@ -31,6 +31,7 @@ _variables:
 versions:
   - dir: '1/debian9/1.15'
     repo: nginx1
+    templateSubDir: nginx
```
Comment (wgrzelak) [inline versions.yaml]: Can you add a warning at the top of the `Dockerfile.template` file?

```
# WARNING: This Dockerfile installs NGINX Prometheus Exporter and should be used
# only in versions 1.14.x and 1.15.x. For newerst version of NGINX, use template
# that doesn't install exporter and use exporter as sidecar. 
# For more information, see exporter directory.
```

File: tests/functional_tests/prometheus_nginx_test.yaml [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+setup:
+- command: [docker, network, create, -d, bridge, testbridge-$UNIQUE-id]
+- command: [docker, run, --privileged=true, --name, some-nginx-$UNIQUE-id, -p, "8080:8080", --net, testbridge-$UNIQUE-id, -d, 'launcher.gcr.io/google/nginx1:latest']
```
Comment (wgrzelak) [inline tests/functional_tests/prometheus_nginx_test.yaml]: Please be consistent and use single quote instead (in whole file).

File: templates/exporter/Dockerfile.template [hunk 11]
```diff
@@ -0,0 +1,26 @@
+FROM {{ .From }}
+{{ $exporter := index .Packages "exporter" }}
```
Comment (wgrzelak) [inline templates/exporter/Dockerfile.template:2]: Consider adding a blank line between to make it more readable.

File: templates/nginx/licences/nginx-prometheus-exporter/LICENSE [hunk 12]
```diff
@@ -0,0 +1,201 @@
+Apache License
```
Comment (wgrzelak) [inline templates/nginx/licences/nginx-prometheus-exporter/LICENSE]: Why do you add Apache License to the nginx itself?

File: templates/exporter/licenses/nginx-prometheus-exporter/LICENSE [hunk 13]
```diff
@@ -0,0 +1,201 @@
+Apache License
```
Comment (wgrzelak) [inline templates/exporter/licenses/nginx-prometheus-exporter/LICENSE:1]: Apache License is not enough, we must to includes licensees of all dependencies that go binary is used.

Please simple copy licenses directory from NGINX.

**Review by wgrzelak — CHANGES_REQUESTED** [review]

File: templates/exporter/licenses/nginx-prometheus-exporter/LICENSE (hunk 13 above)
Comment (armandomiani) [reply templates/exporter/licenses/nginx-prometheus-exporter/LICENSE:1]: Fixed.

File: tests/functional_tests/prometheus_nginx_test.yaml (hunk 10 above)
Comment (armandomiani) [reply tests/functional_tests/prometheus_nginx_test.yaml]: Fixed.

File: templates/exporter/licenses/nginx-prometheus-exporter/LICENSE (hunk 13 above)
Comment (armandomiani) [reply templates/exporter/licenses/nginx-prometheus-exporter/LICENSE:1]: Done.

File: templates/nginx/licences/nginx-prometheus-exporter/LICENSE (hunk 12 above)
Comment (armandomiani) [reply templates/nginx/licences/nginx-prometheus-exporter/LICENSE]: This Apache License is regarding prometheus-exporter which is contained in 1.14 and 1.15 version is already included in master branch, as follows:

https://github.com/GoogleCloudPlatform/nginx-docker/tree/master/1/debian9/1.15/licences/nginx-prometheus-exporter
https://github.com/nginxinc/nginx-prometheus-exporter/blob/master/LICENSE

File: tests/functional_tests/prometheus_nginx_test.yaml (hunk 10 above)
Comment (armandomiani) [reply tests/functional_tests/prometheus_nginx_test.yaml]: Done. Fixed license.

File: versions.yaml (hunk 9 above)
Comment (armandomiani) [reply versions.yaml]: Done.

File: versions.yaml (hunk 8 above)
Comment (armandomiani) [reply versions.yaml]: Done

File: templates/exporter/Dockerfile.template (hunk 11 above)
Comment (armandomiani) [reply templates/exporter/Dockerfile.template:2]: Done.

File: 1/README.md [hunk 14]
```diff
@@ -2,6 +2,12 @@
 
 This image contains an installation Nginx 1.x.
 
+> WARNING: This Dockerfile installs NGINX Prometheus Exporter and should be used
```
Comment (wgrzelak) [inline 1/README.md]: This must be generated from README.yaml. I think you can simple drop it, since Dockerfile contains this note already.

**Review by wgrzelak — APPROVED** [review]

**Pull request merged**